sky/backends/backend_utils.py

```diff
@@ -1767,40 +1765,61 @@ def _update_cluster_status_no_lock(
         return record
 
     cluster_name = handle.cluster_name
-    try:
-        # TODO(zhwu): This function cannot distinguish transient network error
-        # in ray's get IPs vs. ray runtime failing.
-        external_ips = handle.external_ips(use_cached_ips=False)
-        # This happens to a stopped TPU VM as we use gcloud to query the IP.
-        if external_ips is None or len(external_ips) == 0:
-            raise exceptions.FetchIPError(
-                reason=exceptions.FetchIPError.Reason.HEAD)
-        if handle.launched_nodes == 1:
-            # Check the ray cluster status. We have to check it for single node
-            # case, since the get_node_ips() does not require ray cluster to be
-            # running.
-            ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml)
-            runner = command_runner.SSHCommandRunner(external_ips[0],
-                                                     **ssh_credentials)
-            returncode = runner.run('ray status', stream_logs=False)
-            if returncode:
+    use_spot = handle.launched_resources.use_spot
+    # For non-spot clusters, this is an optimization: we call external_ips()
+    # and/or SSH into the cluster to run 'ray status' in various cases to
+    # determine cluster health, because these may be faster than querying
+    # the true node statuses from the cloud provider.
+    # For spot clusters, the above can be unsafe.
+    # Therefore we directly query the cloud provider.
+    if not use_spot:
```

Comment:
We should check the `ray status` on the spot clusters as well. It is possible that the ray cluster is not running, but the cloud provider shows UP, when the user call `ray stop` manually. Our job queue depends on the healthiness of the ray cluster, if the ray cluster is not running, we should consider the cluster as `INIT`.

Reply:
That is to say, we previously assume `ray cluster is healthy` indicates `cloud provider shows UP`, but based on your assumption both should be checked for the spot cluster.

Reply:
that's a good point. I'm looking at the behavior of the current codebase. It seems to me if ray cluster is not running. we will also determine the cluster status by just querying the cloud provider. and if it's running, we will set the state to `UP`. 
https://github.com/skypilot-org/skypilot/blob/35d017dc5fb7cf35ff3b04869a6a6bfda65f4259/sky/backends/backend_utils.py#L1786-L1788

Do you mean we should also set `INIT` for on-demand clusters if ray is not running?

Reply:
No, we will never set the state to `UP` after failing to get the IPs. The code will go through the following code path and set the cluster to `INIT`. https://github.com/skypilot-org/skypilot/blob/1696cce5b1cd3d9edcacecb9cb2d1af3853d9e51/sky/backends/backend_utils.py#L1856-L1890

Reply:
ah sorry I misread the code. then this makes sense. let me add this case to the change.

Reply:
I modified the code to also check ray cluster healthiness for Spot clusters. Now only when `ray_cluster_up` and all clusters are running, we set the state `UP`. Does this look okay?